server_handlers.go

```diff
@@ -83,17 +84,20 @@ func serverHandshakeHandler(c *Conn) (*alert, error) {
 			if err := verifyCertificateVerify(plainText, h.hashAlgorithm, h.signature, c.state.remoteCertificate); err != nil {
 				return &alert{alertLevelFatal, alertBadCertificate}, err
 			}
-			verified := false
+			var chains [][]*x509.Certificate
+			var err error
+			var verified bool
 			if !c.insecureSkipVerify && c.clientAuth >= VerifyClientCertIfGiven {
-				if err := verifyClientCert(c.state.remoteCertificate, c.rootCAs); err != nil {
+				if chains, err = verifyClientCert(c.state.remoteCertificate, c.clientCAs); err != nil {
 					return &alert{alertLevelFatal, alertBadCertificate}, err
 				}
 				verified = true
 			}
 			if c.verifyPeerCertificate != nil {
-				if err := c.verifyPeerCertificate(c.state.remoteCertificate, verified); err != nil {
+				if err := c.verifyPeerCertificate(c.state.remoteCertificate, chains); err != nil {
 					return &alert{alertLevelFatal, alertBadCertificate}, err
 				}
+				verified = true
```

Comment:
@jkralik did this never work/worth adding a test? I can pull it up, not familiar with this as much anymore but this sticks out to me.

Reply:
I look at and check it how is verification is implemented in `crypto/tls` and i found `verified` must be set only when verification passed by our library(not by custom function). Also I removed check of `InsecureSkipVerify` from the server because it is a client option.

 I added tests to cover use cases of verification for server.